Move multiple items - add UI

jimklimov opened the pull request:
Follows up from (and includes) PRs #16 and #17, this one just adds an UI pane allowing to bulk-move the queue items by regex easily (without copying a button URL and editing the link and pasting into browser and going from GET to POST by the new page... too cumbersome).

Getting here took more time than I hoped, due to battle with Jelly and Javascript, and too many typos to count but hard to catch. Result is relatively ugly (suggestions welcome) as there is a large blank space between pane title and table row for textarea and buttons, but finally functionally it works:

![image](https://github.com/user-attachments/assets/47b610b8-ed3c-4fed-85a4-b362fd87e436)

Some technical notes:
* I had to make this a separate pane, so it is not auto-refreshed by AJAX. Otherwise it was nearly impossible to edit the textbox - it was refreshing and emptied every second or two.
  * It would be cool to not-show this pane when there are fewer than 2 or 3 builds queued. Perhaps some javascript in the AJAX part to wholly hide or `toggleCollapse` this pane could help, but is out of my current ability (sadly the experiment took too much time already).
  * I added the pane *above* the queue list, because the latter can be very long (hundreds of items in my use-case) so it takes a while to load and to scroll the browser exactly to the point where the Build Executors are (so finding and using this new pane would be similarly problematic).
* I made an `<f:entry>` around several table data cells in the row, assuming this can help the textbox and TWO buttons that can POST its contents see each other. Maybe this is overkill, did not check further - does work this way at least.
* The original use of `<f:link>` in the list of queue items was not usable here, because server-side pre-generated URL does not really fit. After a lot of experiments, I settled on raw `<button>` as suggested by some comments in e.g. https://github.com/jenkinsci/jenkins/blob/838cafb8fb86fd843b85d2e22e957e09a1ce86a9/war/src/main/webapp/scripts/hudson-behavior.js#L919 `makeButton` and snatched `onClick` code from https://github.com/jenkinsci/jenkins/blob/838cafb8fb86fd843b85d2e22e957e09a1ce86a9/war/src/main/webapp/scripts/hudson-behavior.js#L2566 `validateButton` to POST what we need. The button wraps an `<img>` tag (generated from `<l:icon>` as it was) so the buttons are seen as such but adorned by the image.
* I considered keeping the textbox empty and adding `~/` and `/` in `onClick` code, but keeping it pre-populated allows users to add `i` for case-insensitive match in the end. Also this reminds that the text should be a regex :)

jimklimov committed:
MoveActionWorker: extend API to handle not only movement of a single item, but also an array of those

...add support for regex pattern passed in ITEM_ID_PARAM_NAME
(tried if no match as numeric queue ID or exact job name)

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
docs_src/docs/CLI.md: mention regex mode for bulk queue item movement

...also clarify the docs a bit (much more is needed to make sense of
them though).

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
MoveActionWorker.java, CLI.md: limit regex mode to itemId=~/.../ markup

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
Apply code style fix (spacing for colons and elvis)

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
MoveActionWorker: update method comments

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
MoveActionWorker: fix getItemsBefore() variant for array of itemsA[]

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
Add tests for MoveActionWorker method variants for bulk moves

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
MoveActionWorker: moveImpl(): allow requesting a case-insensitive regex matcher

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
MoveActionWorker: moveDown()/moveUp() variants for bulk move - set desires for all queue items above/below those in itemsA[] with a cut-off at the edge item (if any) that was just above/below the extreme itemA[] element

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
MoveAction_doMoveByNameTest: refactor with a cached queue variable

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
MoveAction_doMoveByNameTest: add a test to doMoveByNameManyByRegex() and moveUp() several items by one, not to top

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
SimpleQueueSorter: sortBuildableItems(): comment practical experience from debug stepping

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
MoveAction_putAOnTopOfBTest: add tests for effective no-ops and for bulk moves of an array of queue items over some entry

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
MoveActionWorker: fix bulk-move variants of move(Up|Down|ToTop|ToBottom) to better see intermediate items between those in itemsA[] set to move

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
MoveActionWorker: bulk-variants of getItemsBefore() and getItemsAfter() are currently not used - current bulk code is better served by getItemsBeforeLast() and getItemsAfterFirst(), but they may be used later, so we just suppress warnings about non-usage

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
SimpleQueueWidget/index.jelly, SimpleQueueUpdateAction/ajaxBuildQueue.jelly: add a non-AJAX pane for bulk move of queue items

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov commented:
Thinking of it, there may also be a fair cause to add a cancel button, which would move remaining matching queued items to top and issue job cancellations. In my practice, teardown of the hundreds of stages can take a few minutes (probably with workspace deletions etc.), and they do try to get a build node needlessly (if they have priority to do so) while the more-promising builds could get handled instead.

jimklimov commented:
Thanks to discussion at Jenkins Summit, de-facto the gap can be removed by adding `style="--section-padding:0;"` to the second pane with "content" for the regex edit area:

<img width="421" height="593" alt="image" src="https://github.com/user-attachments/assets/313a2f06-d3a2-47a5-8b7d-968f47c9df5e" />

or more to the pin-point, changing the first div inside suffices:
<img width="525" height="653" alt="image" src="https://github.com/user-attachments/assets/4aa7d9d5-7f4f-4c1e-9cab-50ec97723453" />

Probably a cleaner solution would be to add an additional CSS class anchored to the `div id` (and adding the `id` for that second area). Alas, this is hidden inside `l:pane` jelly tag, so maybe some Javascript post-processing of the loaded page is needed...

Note the `section-padding` should not be neutered page-wide, it is used for reasonable design in unrelated parts of the "left column".

judovana commented:
hi! Looks good. Am not sure how I could miss this PR. Will look more closer, try and probably merge

jimklimov committed:
Merge remote-tracking branch 'upstream/master' into move-multiple-items

Adjust annotation change from @Nonnull to @NonNull

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
Merge remote-tracking branch 'upstream/master' into move-multiple-items

Adjust annotation change from @Nonnull to @NonNull

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
Merge branch 'move-multiple-items' into move-multiple-items-UI

Including refactors started by @mawinter69

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
Merge branch 'move-multiple-items' into move-multiple-items-UI

Now with fixed imports and suppressed warnings

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov commented:
Nominally updated to current code base foundations, but probably not the latest iteration before merge:
* There are review comments to #17 which may end up in some more fixes to the Java part (so this one may have to be rebased/remerged to minimize the diff after that one does get merged)
* This PR originally changed two `*.jelly` files (identically) and now one, but there is inline JavaScript that @mawinter69 removed elsewhere with PR #24, and probably this PR would need some similar changes for buttons to still work albeit with some new mechanism.

mawinter69 commented on src/main/resources/cz/mendelu/xotradov/SimpleQueueWidget/index.jelly:
```diff
@@ -12,6 +12,33 @@
+                    </td>
+                    <td class="pane" align="center" valign="middle">
+                        <!-- Inspired by validateButton() at https://github.com/jenkinsci/jenkins/blob/838cafb8fb86fd843b85d2e22e957e09a1ce86a9/war/src/main/webapp/scripts/hudson-behavior.js#L2581 -->
+                        <button onClick="var parameters = {}; parameters['${it.getMoveTypeName()}']='UP_FAST'; parameters['${it.getItemIdName()}']=findPreviousFormItem(this, 'BulkMoveRegex').value; fetch('${rootURL}/simpleMove/move', { method: 'post', body: objectToUrlFormEncoded(parameters), headers: crumb.wrap({'Content-Type': 'application/x-www-form-urlencoded',})})" >
```
Using  `onClick` is inline javascript which is not allow when you want to be CSP compliant

jimklimov replied:
I agree in principle, hence the CC :) But at the moment I do not know what to do with this information (e.g. how to change the code so it is better but still works as it did). Probably there's something in PR #24, but its changeset is huge (with the Nonnull changes etc.) so I did not find a solution quickly.

mawinter69 replied:
You will just need to put the code in a js file, load it with an `adjunct` and in the code select the buttons (via id maybe) and attach the onclick listeners to them

jimklimov replied:
Seems to work, just a day of experimentation. Loses the variables known in Jelly so had to hard-code stuff. Not sure if I can pass any arguments into `st:adjunct` for Javascript to interpret as its own, probably not (given that it is same call for half a dozen file types).

mawinter69 commented on src/main/resources/cz/mendelu/xotradov/SimpleQueueWidget/index.jelly:
```diff
@@ -12,6 +12,33 @@
+    <j:if test="${h.hasPermission(app.MANAGE)}">
+        <!-- Keep the textbox out of AJAX pane so it is not erased every couple of seconds -->
+        <l:pane title="Build Queue Bulk Move (by regex)" width="2" id="buildSimpleQueueBulkMove">
+            <tr>
```
Wouldn't it be better to put that inside the existing pane for the queue (line 43).  Though that doesn't allow to make clear that this is for bulk moves, but maybe do that with a placeholder in the input field

judovana replied:
I was just going to suggest to move it to completely independent page,  with full control over queue. But do not have strong preference at the end.  Am only sometimes worried how full the main page can be.  Especially now when many buttons were moved to the gear wheel at left up corener, may be undesired to see more here.

jimklimov replied:
Not sure if the plugin behavior changed with your recent PRs. With the previous releases' baseline, the big section with the list of queue items is auto-refreshed once in a while. Having the editable text box (for name/regex) in that pane had it re-rendered and any edits forgotten. Having it outside allows it to remember the last-used pattern for much longer (sometimes even across mobile browser restart); a full-page refresh clears it though.

mawinter69 replied:
yeah the reload would be a problem

mawinter69 replied:
One idea might be to just have a button and on click a dialog opens where you can enter the regex.

judovana replied:
> One idea might be to just have a button and on click a dialog opens where you can enter the regex.

Rather then that I would rather see an custom backend, where may appear eg comboboxes like "search by text", "search by regexe, later  "search by id" or "by full name". Similarly there may be butons to affect how to move. Up? Down? View? and so on...

judovana replied:
Forgotten note - I'm completely ok with current approach of the small textfield above queue

jimklimov replied:
So far worked around this eventually (setting a magic variable in the new `ajax.jelly` and checking against it in code).

jimklimov replied:
Later also dropped the magic variable to use common `ajax==null` that for some reason did not work for me the first time around (maybe flawed experiment) but did later.

jimklimov committed:
Merge remote-tracking branch 'upstream/master' into move-multiple-items

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
MoveActionWorker: fix back proper @SuppressFBWarnings usage

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
Merge branch 'move-multiple-items' into move-multiple-items-UI

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov commented:
Taking baby steps to update UI part of the PR.

With a build from the current commit (faa016b) it looks odd, adding the search area header twice (disregard the un-rendered buttons, file names not changed yet) while some jobs run but (I think) there are no items in queue anymore:
<img width="369" height="479" alt="image" src="https://github.com/user-attachments/assets/be4d9696-f6d0-4fb0-aed6-6571c031df69" />

When I refresh the page with no builds running, it first looks alright and after some 5 seconds changes, becoming simliar to what I posted above. It does not refresh further (text I typed into both boxes remained there for over a minute).
<img width="339" height="367" alt="image" src="https://github.com/user-attachments/assets/58e50b56-2a23-42ae-a51d-127f4f576155" /> <img width="353" height="401" alt="image" src="https://github.com/user-attachments/assets/ec2dede6-3842-419c-ba4d-61d1d7a15502" />

Real queue looks better though:
<img width="349" height="577" alt="image" src="https://github.com/user-attachments/assets/d77023c5-7d37-48bc-b736-592092200a76" />

jimklimov committed:
SimpleQueueWidget/index.jelly: use new SVG button images for buildSimpleQueueBulkMove pane

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
SimpleQueueWidget/index.jelly: move javascript for BulkMoveRegex buttons into a static file, loaded via st:adjunct (for CSP compliance)

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
SimpleQueueWidget/index.jelly: remove padding around BulkMoveRegex textbox

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov commented:
After updates to relocate javascript and use new buttons:

* No builds, no queue - no regex pane when not logged in (permissions check):
<img width="354" height="258" alt="image" src="https://github.com/user-attachments/assets/12264451-ac25-46e1-b008-0bd6357768b0" />

* After login: regex pane on top appears (expected), empty queue pane becomes another regex pane after a few seconds (not expected):
<img width="361" height="388" alt="image" src="https://github.com/user-attachments/assets/f6ae4e05-0b1c-4dc9-ac1e-0420f1f027c8" /> <img width="348" height="379" alt="image" src="https://github.com/user-attachments/assets/539269d7-d3c1-4ef4-b37c-0dedde1b60bb" />

* After scheduling a series of builds, the queue is seen for a few seconds... then is replaced by the regex pane (in whole, no queue list - just the actual running build numbers under Build Executor Status flicker):
<img width="348" height="548" alt="image" src="https://github.com/user-attachments/assets/db0eace6-095c-4a9a-b244-af4e947d9594" />

* Trivial test pipeline:
```
pipeline {
    agent any
    stages {
        stage ("Discovery") {
            steps {
                sleep 10
            }
        }
    }
}
```

I click "Build" a dozen times and refresh the dashboard for UI test evaluation above.

judovana commented:
Long time ago, this extension had a bad habit to disappear after few seconds of being visible...  Hopefully it was not reintoruced by recent changes and/or by this PR. I'm afraid my jelly skills are to low to have any reasonable advice. @mawinter69 , do you?

jimklimov committed:
SimpleQueueWidget/index.jelly, ajax.jelly: hack around loading regex pane as the queue pane after auto-refresh

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
SimpleQueueWidget/index.jelly, bulk-move-buttons.js: fix externalized JS fixup of bulk move buttons onclick action

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
SimpleQueueWidget/index.jelly, bulk-move-buttons.js, simple-queue-widget.css: fix style of the Bulk Move area as JS post-processing

NOTE: This also reduces blank space between Bulk Move box and actual
Build Queue box, arguably a good thing as they are related.

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

mawinter69 commented on src/main/resources/cz/mendelu/xotradov/SimpleQueueWidget/ajax.jelly:
```diff
@@ -1,6 +1,7 @@
 <?jelly escape-by-default='true'?>
 <j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:l="/lib/layout">
   <l:ajax>
+    <j:set var="ajax_mode" value="true"/>
```
`l:ajax` already sets a variable with name `ajax` so you can delete this line and use it below

jimklimov replied:
I tried, but did not find what value it sets. Checks for `ajax==null` or `h.isAutoRefresh(request)` blocked this pane from appearing at all, and for other values I've randomly tried, it kept appearing instead of the queue eventually.

jimklimov replied:
I'll double-check now, just in case I did some of those tests without logging in. (After the WAR restart, my local Jenkins sometimes wants new login sessions and sometimes remembers them, it seems.)

jimklimov replied:
My bad I guess, works now. Shame on me, lost my time :D

mawinter69 commented on src/main/resources/cz/mendelu/xotradov/SimpleQueueWidget/index.jelly:
```diff
@@ -12,6 +12,35 @@
     </j:choose>
+
+    <!-- ajax_mode is set in ajax.jelly which re-includes this page -->
+    <j:if test="${ajax_mode!='true' and h.hasPermission(app.MANAGE)}">
```
```suggestion
    <j:if test="${ajax==null and h.hasPermission(app.MANAGE)}">
```

See the end of this file where the same pattern is used to avoid that this part is realoaded in the ajax call

jimklimov commented:
With recent commits, the local build of the plugin seems to restore features I am after:
* Build Queue only shows the queue (or lack thereof) and does auto-refresh every few seconds
* Bulk Move pane is rendered once
** Now it also looks nicer with style post-applied via JavaScript to remove the blank padding area
* Bulk Move buttons work via externalized JavaScript (some loss of code fidelity to hard-coding as jelly variables for `${it.getMoveTypeName()}`, `${it.getItemIdName()}`, `${rootURL}` can not be used anymore)
** For the rootURL used `document.head.dataset.rooturl` as it seems many plugins do so; not sure this is correct for Jenkins instances not at the root of their server hostname though.

Screenshots:

* No builds (logged in):
<img width="353" height="326" alt="image" src="https://github.com/user-attachments/assets/554f9309-f488-4d0c-a62e-ee49a5636ad1" />

* Builds just started (in light-weight processing before getting queued to a node):
<img width="356" height="624" alt="image" src="https://github.com/user-attachments/assets/fca9de0d-fe12-41c1-9296-910cd5453c17" />

* Builds waiting for the node/executor:
<img width="351" height="493" alt="image" src="https://github.com/user-attachments/assets/f865932f-c309-41bc-a6c3-5ac66f5b5c70" />

Clicks into the sorting buttons of individual items in the queue instantly cause the full page refresh. This is what the plugin did before my modifications, and largely caused them: moving dozens or hundreds of items this way was untenable, even if the page became laggy and I could click several arrows => post the HTTP requests before the page actually began re-rendering.

mawinter69 commented on src/main/resources/cz/mendelu/xotradov/SimpleQueueWidget/bulk-move-buttons.js:
```diff
@@ -0,0 +1,21 @@
+        parameters['itemId']=findPreviousFormItem(this, 'BulkMoveRegex').value;
+
+        // Tried to use jenkins.baseUrl provided by jenkins.js but it is not global or something?
+        fetch(document.head.dataset.rooturl + '/simpleMove/move', {
```
This is correct. When jenkins runs under context this contains the context url, e.g. `/jenkins` when you run with `mvn hpi:run`

jimklimov replied:
Thanks

mawinter69 commented on src/main/resources/cz/mendelu/xotradov/SimpleQueueWidget/index.jelly:
```diff
@@ -12,6 +12,35 @@
+                        <f:textbox name="BulkMoveRegex" value="~//"/>
+                    </td>
+                    <td class="pane" align="center" valign="middle">
+                        <button id="BulkMoveRegex_button_up_fast" class="BulkMoveRegex_button" data-var="UP_FAST">
```
Please add `jenkins-button` to the class so the button looks like other Jenkins buttons

jimklimov replied:
done

mawinter69 commented on src/main/resources/cz/mendelu/xotradov/SimpleQueueWidget/index.jelly:
```diff
@@ -12,6 +12,35 @@
+                        </button>
+                    </td>
+                    <td class="pane" align="center" valign="middle">
+                        <button id="BulkMoveRegex_button_down_fast" class="BulkMoveRegex_button" data-var="DOWN_FAST">
```
same here

jimklimov replied:
done

jimklimov committed:
SimpleQueueWidget/index.jelly: add jenkins-button class to Bulk Move elements

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov commented:
With added style, buttons are even bigger:
<img width="345" height="461" alt="image" src="https://github.com/user-attachments/assets/2c0f8d7f-1ce6-41ea-b592-5518bee499df" />
Probably easier to tap, but eating away almost half the Bulk Move area...

jimklimov committed:
SimpleQueueWidget/index.jelly, ajax.jelly: use standard ajax==null check after all

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

mawinter69 commented on src/main/resources/cz/mendelu/xotradov/SimpleQueueWidget/index.jelly:
```diff
@@ -12,6 +12,34 @@
+        <l:pane title="Build Queue Bulk Move (by regex)" width="2" id="buildSimpleQueueBulkMove">
+            <tr>
+                <f:entry>
+                    <!-- The first TD is where stage/job/part names appear normally -->
```
Remove that surrounding `f:entry` this causes a gap above the input field for me

jimklimov replied:
Aha... maybe that's the same one as what I had in earlier screenshots and valiantly fixed today with post-processed styles? Thanks...

jimklimov replied:
That was it! :)

With both `f:entry` and style-hack removed:
<img width="362" height="327" alt="image" src="https://github.com/user-attachments/assets/e181dbaa-6d35-4981-afbe-e242af0de1f3" />

jimklimov committed:
SimpleQueueWidget/index.jelly, CSS, JS: remove f:entry and markup-padding problems go away

Thanks @mawinter69 !

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov commented:
With `f:entry` and style hacks both removed, the layout now looks like this:

<img width="362" height="327" alt="image" src="https://github.com/user-attachments/assets/e181dbaa-6d35-4981-afbe-e242af0de1f3" />

mawinter69 commented:
> With added style, buttons are even bigger: <img alt="image" width="345" height="461" src="https://private-user-images.githubusercontent.com/1636756/561011165-2c0f8d7f-1ce6-41ea-b592-5518bee499df.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzMxNTU5NjgsIm5iZiI6MTc3MzE1NTY2OCwicGF0aCI6Ii8xNjM2NzU2LzU2MTAxMTE2NS0yYzBmOGQ3Zi0xY2U2LTQxZWEtYjU5Mi01NTE4YmVlNDk5ZGYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDMxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAzMTBUMTUxNDI4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTFhZDY2ZTM1Y2I5OGQzNTE4YjNjMDhmNzRmMTZjMzliYWUzM2Y1NTM0NTk5ZTI1NTUxZTE0ZTI2ODllODkwYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.BXzcv4GDvqIZijgQ1C2Jt_uFtzxsgWHFzcSCdvnOhJM"> Probably easier to tap, but eating away almost half the Bulk Move area...

Could add some css to make them smaller 
<img width="345" height="120" alt="image" src="https://github.com/user-attachments/assets/efb69a4a-e17a-4548-90cd-1027c6cfeb19" />

jimklimov commented:
Oh please indulge me: which style incantation would that be? :D

mawinter69 commented:
> Oh please indulge me: which style incantation would that be? :D

This will produce square buttons:
```
.BulkMoveRegex_button {
    padding: 2px 10px;
}
```
My sample above used  `padding: 2px 6px;`

jimklimov committed:
SimpleQueueWidget/simple-queue-widget.css: make the Bulk Move buttons smaller

Co-authored-by: Marcus Winter <m.winter@sap.com>
Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
MoveActionWorker.java: extend findItemsByPattern() to support Matcher::(find|matches) logic, and in moveImpl() cascade from exact string to trying regex as Matcher::matches (if not ~/.../ formatted)

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
Merge branch 'move-multiple-items' into move-multiple-items-UI

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
MoveAction_doMoveByNameTest: add tests to differentiate Matcher::matches() vs. Matcher::find()

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
MoveActionWorker.java, MoveAction_doMoveByNameTest.java, SimpleQueueWidget.java: introduce "itemIdMode" parameter ("regex" or not)

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
Merge branch 'move-multiple-items' into move-multiple-items-UI

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
SimpleQueueWidget/bulk-move-buttons.js: set itemIdMode="regex"

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov commented:
Updated to match changes requested in PR #17, not tested locally yet (maybe in a few hours).

jimklimov commented:
Now tested interactively, worked for both `~/.5./` (e.g. `part of ... #151`) and `.*5.*` (e.g. `part of ... #145` and `part of ... #151`) job numbers, as well as `.*161` (`part of ... #161`), to sort them up and down (top/bottom of the queue).

As expected (no regression added) did not work for plain `161` as no display name of a job part matched that string exactly.
In fact, as a number that would be interpreted as a queue item number (and this Jenkins uptime did not have that many builds yet). Re-checked that `85` (in the screenshot below) for `part of ... #176` indeed got that line sorted upwards.

<img width="520" height="414" alt="image" src="https://github.com/user-attachments/assets/3e75d6a5-114e-44bc-aa24-c02486f738de" />

jimklimov committed:
README.md: revise wording, fix layout to a more native use of markdown

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CLI.md: drop duplicate query parameter explanations in Legacy API chapter

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CLI.md: rephrase the note about HTTP-302 return value

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CLI.md: revise wording, fix layout to a more native use of markdown (and mostly fit in 120 columns)

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CLI.md: fix curl examples to use scripting-friendly (and Jenkins-common) envvar ${JENKINS_URL}

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CLI.md: add a disclaimer about "job name" == "Queue Task Display Name"

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CLI.md: introduce itemIdMode option, detail regex nuances as a separate chapter

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CLI.md: group troublesome items as "caveats"

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
fixup! CLI.md: revise wording, fix layout to a more native use of markdown (and mostly fit in 120 columns)

jimklimov committed:
MoveActionWorker, CLI.md: now that itemIdMode is explicit, do not check for regex *after* trying queue id (number) or exact string hits, but rather *instead* [#17]

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
MoveAction_doMoveByNameTest: comment the doMoveByName() logic similarly to other test cases

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
MoveAction_doMoveTest: minimize the structural diff vs. MoveAction_doMoveByNameTest::doMoveByName() and rename to make the logical difference clearer

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
Merge branch 'move-multiple-items' into move-multiple-items-UI

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
CLI.md: update that UI for regex mode is now available [#18]

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
SimpleQueueWidget/index.jelly: add tooltips to BulkMoveRegex buttons

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
README.md: add queue_bulk_move_screenshot.png

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov commented:
Updated against PR #17 again, revised docs to mention that UI for regex sorting now exists and added a screenshot.

jimklimov committed:
README.md: relocate "Jenkins Version" badge to start of string so it renders; bump the version to match pom.xml

Signed-off-by: Jim Klimov <jimklimov+jenkinsci@gmail.com>

jimklimov committed:
Merge branch 'master' into move-multiple-items-UI

jimklimov commented:
Ultimately, the remaining changes (compared to #17) brought by this PR are about the jelly pane to enable Bulk Move of items via UI, CSS/JS resources for that, and a documentation bump.

@judovana @mawinter69 : can this be merged too, please? Then I can finally stop building my fork to reap the benefits on my CI servers :D

judovana commented:
Actually why not.  Only I will get rid of type=regex before release. ook?

jimklimov commented:
Why? To move by numeric queue ID or exact full display names? That makes no sense :)
You guys asked to make it an explicit option, and the Bulk Move pane is to move hundreds of items, not one by one (existing arrow buttons in the queue list do that already).

Or is there some misunderstanding again? ;)

jimklimov commented:
Looking at https://github.com/jenkinsci/simple-queue-plugin/pull/28/changes I see that jelly/JS `data-something-whatever` syntax can be used to pass more than one variable (determined by Jelly), so where I complained about hard-coding some strings that `${it.getMoveTypeName()}` instead of `moveType` etc. could convey in pure jelly but no longer does here - this feature can be restored. Maybe in a follow-up PR though, not sure I'll have time for this today. And it's a no-op change as far as everyone is concerned :)

judovana commented:
> Why? To move by numeric queue ID or exact full display names? That makes no sense :) You guys asked to make it an explicit option, and the Bulk Move pane is to move hundreds of items, not one by one (existing arrow buttons in the queue list do that already).

I can actually imagine all three cases.   And I did not  want the type param in the url.  I wanted the waterfall, in old  api so it remains flexible, and new, second exact api where the type will be mandatory. Also I asked `/i` to be reusable by "java style". It will delay the release just for a day. Am lookign froward to have those features out asap too.

jimklimov commented:
> I can actually imagine all three cases.

Well, with the `itemIdMode=regex` option, probably copy-pasting a name with weird characters into the search box would not work anymore (I can imagine the JS `objectToUrlFormEncoded()` could ensure that earlier, but did not check).

Ultimately, for the release - if the Bulk Move pane remains useful for my interactive use-cases, entering the momentarily most-useful information with minimal typing and clicking/tapping (e.g. type `1234` maybe into the right spot of the pre-filled text area, and press the button to move a few dozen hits), I am okay with any under-the-hood changes you need to make. Now this is your code base :)

jimklimov commented:
> Also I asked `/i` to be reusable by "java style". 

I think it still is: if the `idParam` does not start with `~/` the regex logic goes to "Java style" mode, and passes the whole string to `Matcher::matches`. So if *that* implementation has special handling for the string ending with `/i` or any other special flags - it would be handled. I wonder if for that it should also start with a slash though? Anyhow, as long as it is not `~/` but just `/` in the beginning, it goes to Java style...

jimklimov commented:
Side question: is there some separate docs publication activity? I see that https://jenkinsci.github.io/simple-queue-plugin/CLI/ did not update after merging now several PRs that changed `CLI.md`. Is this tied to a new release, or triggered manually?

judovana commented:
> > I can actually imagine all three cases.
> 
> Well, with the `itemIdMode=regex` option, probably copy-pasting a name with weird characters into the search box would not work anymore (I can imagine the JS `objectToUrlFormEncoded()` could ensure that earlier, but did not check).
> 

The `objectToUrlFormEncoded` or similar is hanging in my mind.  I have couple of jobs contain "~" so that is somethign I have to try for both regex end exact match. 

> Ultimately, for the release - if the Bulk Move pane remains useful for my interactive use-cases, entering the momentarily most-useful information with minimal typing and clicking/tapping (e.g. type `1234` maybe into the right spot of the pre-filled text area, and press the button), I am okay with any under-the-hood changes you need to make. Now this is your code base :)

Right. That is the goal. I will submit a PR, and will kindly ask you to try it out if possible. No more delays planed. Your usecase should be carved in stone.. argh.. to unittests :) (which are more solid then my memory). ty!

judovana commented:
> Side question: is there some separate docs publication activity? I see that https://jenkinsci.github.io/simple-queue-plugin/CLI/ did not update after merging now several PRs that changed `CLI.md`. Is this tied to a new release, or triggered manually?

IIRC, that will update after release. If not, will be elaborated on.

jimklimov commented:
Thanks!

judovana commented:
> > Ultimately, for the release - if the Bulk Move pane remains useful for my interactive use-cases, entering the momentarily most-useful information with minimal typing and clicking/tapping (e.g. type `1234` maybe into the right spot of the pre-filled text area, and press the button), I am okay with any under-the-hood changes you need to make. Now this is your code base :)
> 
> Right. That is the goal. I will submit a PR, and will kindly ask you to try it out if possible. No more delays planed. Your usecase should be carved in stone.. argh.. to unittests :) (which are more solid then my memory). ty!

@jimklimov  pls eyball: https://github.com/jenkinsci/simple-queue-plugin/pull/29
